pkg/app/server.go

```diff
@@ -87,7 +88,7 @@ func RunKubeStateMetricsWrapper(ctx context.Context, opts *options.Options) erro
 // RunKubeStateMetrics will build and run the kube-state-metrics.
 // Any out-of-tree custom resource metrics could be registered by newing a registry factory
 // which implements customresource.RegistryFactory and pass all factories into this function.
-func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
+func RunKubeStateMetrics(ctx context.Context, opts *options.Options, r *discovery.GVKMaps) error {
```

Comment:
I would expect the discovery logic to be created inside of `RunKubeStateMetrics` as it shouldn't be useful outside of that scope

Reply:
I've moved the logic further down, while keeping the `gvkMaps` object global to the caller, in order to ensure we operate on the preprequisites for the `discovery` mechanism only once. I wanted to do this using `context.WithValues`, but we already have a `context.WithCancel` in place so I passed down this additional param.